### PR TITLE
Refactor Dask cuDF legacy code

### PR DESCRIPTION
**Pull request by rjzamora**
## Description

The "legacy" DataFrame API is now deprecated (https://github.com/dask/dask/pull/11437). The main purpose of this PR is to start isolating legacy code in Dask cuDF.

**Old layout**:

```
dask_cudf/
├── expr/
│   ├── _collection.py
│   ├── _expr.py
│   ├── _groupby.py
├── io/
│   ├── tests/
│   ├── ...
│   ├── parquet.py
│   ├── ...
├── tests/
├── accessors.py
├── backends.py
├── core.py
├── groupby.py
├── sorting.py
```

**New layout**:

```
dask_cudf/
├── _expr/
│   ├── accessors.py
│   ├── collection.py
│   ├── expr.py
│   ├── groupby.py
├── _legacy/
│   ├── io/
│   ├── core.py
│   ├── groupby.py
│   ├── sorting.py
├── io/
│   ├── tests/
│   ├── ...
│   ├── parquet.py
│   ├── ...
├── tests/
├── backends.py
├── core.py
```

**Notes**

- This PR adds some backward compatibility to the expr-based API that was previously missing: The user can now import collection classes from `dask_cudf.core` (previously led to a "silent" bug when query-planning was enabled). 
- The user can also import various IO functions from `dask_cudf.io` (and sub-modules like `dask_cudf.io.parquet`), but they will typically get a deprecation warning.
- This PR is still technically "breaking" in the sense that the user can no longer import *some* functions/classes from `dask_cudf.io.*`. Also, the `groupby`, `sorting`, and `accessors` modules have simply moved. It *should* be uncommon for down-stream code to import from these modules. It's also worth noting that query-planning was already causing problems for these users if they *were* doing this.

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: python/dask_cudf/dask_cudf/__init__.py [hunk 1]
```diff
@@ -2,20 +2,17 @@
 
 from dask import config
 
-# For dask>2024.2.0, we can silence the loud deprecation
-# warning before importing `dask.dataframe` (this won't
-# do anything for dask==2024.2.0)
-config.set({"dataframe.query-planning-warning": False})
```
Comment (rjzamora) [inline python/dask_cudf/dask_cudf/__init__.py:8]: No longer relevant.

File: python/dask_cudf/dask_cudf/_expr.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+        other = new_collection(other).to_backend("pandas")._expr
+
+    # Call "original" function
+    return _original_get_divisions(frame, other, *args, **kwargs)
+
+
+_PATCHED = False
+
+
+def _patch_dask_expr():
```
Comment (rjzamora) [inline python/dask_cudf/dask_cudf/_expr.py:202]: Most of this PR is just moving around existing code. This change is actually new - We used to implicitly "patch" dask-expr when the `expr` module was imported. I'm trying to make the patching a bit more explicit (but I only want to allow this function to do its thing once).

**Comment by madsbk** [comment]
If `_accessors.py` & co. are all private, would it make sense to move them to a sub-folder and avoid the `_` prefix?

**Comment by wence-** [comment]
Question: is there any user-visible change to this refactoring, such that we must warn users about the deprecation? Or is it that we can just swap out the implementation from under them?

**Comment by rjzamora** [comment]
>If _accessors.py & co. are all private, would it make sense to move them to a sub-folder and avoid the _ prefix?

I'm happy to do whatever others prefer. I just want to choose *some* convention that helps us move in the direction of making everything other than `dask_cudf.<function-or-class>` private. [Dask-expr](https://github.com/dask/dask-expr/tree/main/dask_expr) uses a flat directory of `.py` files with the `_` prefix. We can certainly put everything in a `_`-prefixed directory instead.

>Question: is there any user-visible change to this refactoring, such that we must warn users about the deprecation? Or is it that we can just swap out the implementation from under them?

There may be some code that people were importing directly from `dask_cudf.io` (e.g. `dask_cudf.io.parquet.create_metadata_file`). If I understand correctly, everything else is already broken when query-planning is enabled, so we would only cause "new" problems for legacy-API users (and the legacy API will definitely be gone for 25.02).

With that said, I'll try to make some changes to better-warn people who try to import things from "deprecated" paths.

**Comment by madsbk** [comment]
@rjzamora agree, the important thing is to mark them private.
What about something like:
```
dask_cudf/
├── _expr/
│   ├── collection.py
│   ├── expr.py
│   ├── groupby.py
│   ├── accessors.py
```

**Comment by rjzamora** [comment]
>What about something like:

Sounds good @madsbk - I made this change.

File: python/dask_cudf/dask_cudf/__init__.py [hunk 3]
```diff
@@ -38,34 +36,51 @@ def read_parquet(*args, **kwargs):
...
+            module = import_module(".".join(new_attr[:-1]))
+            return getattr(module, new_attr[-1])(*args, **kwargs)
+
+        # No alternative - raise an error
         raise NotImplementedError(
-            f"Top-level {attr_name} API is not available for dask-expr."
+            f"{old_api} is no longer supported. " + (rec or "")
         )
 
     return inner_func
```
Comment (rjzamora) [inline python/dask_cudf/dask_cudf/__init__.py:55]: **Note**: I decided to expand this utility to help deal with the many `dask_cudf.io` APIs that are no longer directly accessible with query-planning on.

File: python/dask_cudf/dask_cudf/__init__.py [hunk 4]
```diff
@@ -38,34 +36,51 @@ def read_parquet(*args, **kwargs):
...
-    from .groupby import groupby_agg  # noqa: F401
-    from .io import read_text, to_orc  # noqa: F401
+    from ._legacy.groupby import groupby_agg  # noqa: F401
+    from ._legacy.io import read_text  # noqa: F401
+    from . import io  # noqa: F401
+
+
+to_orc = _deprecated_api(
+    "dask_cudf.to_orc",
+    new_api="dask_cudf._legacy.io.to_orc",
```
Comment (rjzamora) [inline python/dask_cudf/dask_cudf/__init__.py:74]: The legacy to_orc API is actually "fine" (it's compatible with query-planning stuff), but we might as well discourage this kind of usage.

**Review by madsbk — APPROVED** [review]
Looks good to me, thanks @rjzamora

**Pull request merged**